### PR TITLE
[sw, dif_kmac] Add implementation of KMAC (blocking only)

### PR DESCRIPTION
**Pull request by mundaym**

### CODE REVIEW

**Comment by mundaym** [comment]
@moidx This is still a rough cut and needs more tests but should be mostly working (although I haven't tested non-empty customization strings at all yet). I'm on vacation on Monday but should be able to finish this off on Tuesday.

**Comment by mundaym** [comment]
By the way the desired output length (`l`) is currently in 32-bit words. I don't think I documented that yet.

**Comment by mundaym** [comment]
@moidx I'm not sure what I'm doing wrong but when I try to put the key into share1 I don't get the correct result. For example the following patch causes the test to fail:

```diff
diff --git a/sw/device/tests/dif/dif_kmac_kmac_smoketest.c b/sw/device/tests/dif/dif_kmac_kmac_smoketest.c
index 4a7949ee7..4311f48e9 100644
--- a/sw/device/tests/dif/dif_kmac_kmac_smoketest.c
+++ b/sw/device/tests/dif/dif_kmac_kmac_smoketest.c
@@ -140,9 +140,9 @@ const kmac_test_t kmac_tests[] = {
         .mode = kDifKmacModeKmacLen128,
         .key =
             (dif_kmac_key_t){
-                .share0 = {0x43424140, 0x47464544, 0x4B4A4948, 0x4F4E4D4C,
+                .share1 = {0x43424140, 0x47464544, 0x4B4A4948, 0x4F4E4D4C,
                            0x53525150, 0x57565554, 0x5B5A5958, 0x5F5E5D5C},
-                .share1 = {0},
+                .share0 = {0},
                 .length = kDifKmacKeyLen256,
             },
         .message = "\x00\x01\x02\x03",
```

I'll investigate further tomorrow.

File: sw/device/lib/dif/dif_kmac.h [hunk 1]
```diff
@@ -172,6 +172,12 @@ typedef struct dif_kmac {
    */
   bool squeezing;
 
+  /**
+   * Whether the output length (d) in bits should be right encoded and appended
+   * to the end of the message.
```
Comment (moidx) [inline sw/device/lib/dif/dif_kmac.h:177]: nit: Can you add the modes where this is required?

**Review by moidx — APPROVED** [review]

**Comment by moidx** [comment]
Hi @mundaym, can you open an issue regarding the observed `key_share_0 = {0}` behavior? We can mark the issue as type: Question, Component: RTL. If this is intended behavior, we can update the RTL spec documentation to reflect this.

**Comment by mundaym** [comment]
Thanks @moidx. Opened this issue for the question about the key share behavior: https://github.com/lowRISC/opentitan/issues/6637.

File: sw/device/lib/dif/dif_kmac.h (hunk 1 above)
Comment (mundaym) [reply sw/device/lib/dif/dif_kmac.h]: Done (just KMAC).

**Comment by mundaym** [comment]
@moidx From the discussion in issue #6637 it looks like share 1 is ignored when the EnMasking parameter is set to 0 which it currently is for earlgrey. Do you need masked keys for your work?
